hydra/utils.py

```diff
@@ -53,67 +104,184 @@ def instantiate(config: Any, *args: Any, **kwargs: Any) -> Any:
     """
-
-    if OmegaConf.is_none(config):
+    # Return None if config is None
+    if config is None or OmegaConf.is_none(config):
         return None
 
-    if isinstance(config, TargetConf) and config._target_ == "???":
+    # TargetConf edge case
+    if isinstance(config, TargetConf) and config.is_missing():
         # Specific check to give a good warning about failure to annotate _target_ as a string.
         raise InstantiationException(
             f"Missing value for {type(config).__name__}._target_. Check that it's properly annotated and overridden."
             f"\nA common problem is forgetting to annotate _target_ as a string : '_target_: str = ...'"
         )
 
-    if not (
-        isinstance(config, dict)
-        or OmegaConf.is_config(config)
-        or is_structured_config(config)
-    ):
-        raise HydraException(f"Unsupported config type : {type(config).__name__}")
+    # Structured Config always converted first to OmegaConf
+    if is_structured_config(config):
+        config = OmegaConf.structured(config, flags={"allow_objects": True})
 
+    # Make shallow copy if config is a standard dictionary
     if isinstance(config, dict):
-        configc = config.copy()
-        _convert_container_targets_to_strings(configc)
-        config = configc
-
-    kwargsc = kwargs.copy()
-    _convert_container_targets_to_strings(kwargsc)
-    kwargs = kwargsc
+        config = config.copy()
 
-    # make a copy to ensure we do not change the provided object
-    config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+    # If OmegaConf, copy, make it mutable and set parent for live interpolation
     if OmegaConf.is_config(config):
+        config_copy = OmegaConf.structured(config, flags={"allow_objects": True})
+        OmegaConf.set_readonly(config_copy, False)
+        OmegaConf.set_struct(config_copy, False)
         config_copy._set_parent(config._get_parent())
-    config = config_copy
-    assert OmegaConf.is_config(config)
-    OmegaConf.set_readonly(config, False)
-    OmegaConf.set_struct(config, False)
+        config_copy._set_flag("allow_objects", True)
+        config = config_copy
 
-    target = _get_target_type(config, kwargs)
-    try:
-        config._set_flag("allow_objects", True)
-        final_kwargs = _get_kwargs(config, **kwargs)
-        convert = _pop_convert_mode(final_kwargs)
-        if convert == ConvertMode.PARTIAL:
-            final_kwargs = OmegaConf.to_container(
-                final_kwargs, resolve=True, exclude_structured_configs=True
+    # Finalize config (convert targets to strings, merge with kwargs)
+    # TODO: kwargs are used as a config now. If it contains dataclass
+    # objects, they will be interpreted as config nodes in the
+    # instantiation result. We might want to change that behavior.
+    _convert_container_targets_to_strings(kwargs)
+    _convert_container_targets_to_strings(config)
```

Comment:
Could be a good opportunity to simplify this.
Now that OmegaConf supports objects, we should be able to just convert forward.
instead of converting types to strings (and then locate them again), we can convert everything to types.

Reply:
I agree.

However it seems we can't really store types in OmegaConf containers.

```python
from omegaconf import OmegaConf

class MyClass:
    pass

OmegaConf.create({"_target_": MyClass})
```

raises

```bash
UnsupportedValueType: Value 'MyClass' is not a supported primitive type
	full_key: _target_
	reference_type=Optional[Dict[Union[str, Enum], Any]]
	object_type=dict
```

Reply:
actually my bad, I forgot the flag {"allow_objects"} so it works as expected

Reply:
There is some ambiguity with dataclasses though

```python
@dataclass
class MyClass:
    x: int


cfg = OmegaConf.create({"_target_": MyClass}, flags={"allow_objects": True})
print(cfg._target_)  # {'x': '???'}
```

Reply:
That again, sounds like another case fox boxing the object.